README.md
Evented communication with RFXtrx433.
=====================================

How to Use
==========

rfxcom depends on the serialport module.

To install
----------

<pre>
  npm install rfxcom
</pre>

This will pull in the two dependencies, serialport 1.0.6+ and underscore.js.

To Use
------

<pre>
var rfxcom = require('./rfxcom'),
    pg = require('pg').native,
    conString = "pg://user:password@localhost/user",
    client = new pg.Client(conString);

var rfxtrx = new rfxcom.RfxCom("/dev/ttyUSB0", {debug: true});

/*
 * This reports security updates from X10 security devices.
 */
rfxtrx.on("security1", function (evt) {
  if (evt.deviceStatus === rfxcom.security.MOTION) {
    console.log("Device %s %s detected motion.", evt.subtype, evt.id);
  } else if (evt.deviceStatus === rfxcom.security.NOMOTION) {
    console.log("Device %s %s reported motion stopped.", evt.subtype, evt.id);
  }
});

rfxtrx.on("elec2", function (evt) {
  // Requires a PostgreSQL table
  // CREATE TABLE energy (recorded_time timestamp DEFAULT NOW(),
  //                      device_id VARCHAR, current_watts FLOAT)
  client.query("INSERT INTO energy(device_id, current_watts) values($1, $2)",
                [evt.id, evt.currentWatts]);
});

rfxtrx.initialise(function () {
    console.log("Device initialised");
});
</pre>

LightwaveRf
-----------
There's a specialised Lighting5 prototype, which uses an RfxCom object.

<pre>
    var rfxtrx = new rfxcom.RfxCom("/dev/ttyUSB0", {debug: true}),
        lightwaverf = new rfxcom.Lighting5(rfxtrx, rfxcom.Lighting5.LIGHTWAVERF);

    lightwaverf.switchOn("0xF09AC8/1", {command: rfxcom.Lighting5.MOOD1});
    lightwaverf.switchOn("0xF09AC8/2", {level: 80});
</pre>

I've tested it with both LightwaveRf lights, and the relay switch.

LightwaveRf lights get their identity from the remote used to pair, if you
don't have a remote, or if you want to specify the address manually, you can
put the device into pairing mode and send lightwaverf.switchOn("<address>"),
to unpair, send lightwave.switchOff("<address>") while the device is in pairing
mode.

Lighting2
---------
There's a specialised Lighting2 prototype, which uses an RfxCom object.

<pre>
    var rfxtrx = new rfxcom.RfxCom("/dev/ttyUSB0", {debug: true}),
        lighting2 = new rfxcom.Lighting2(rfxtrx, rfxcom.Lighting2.HOMEEASY_EU);

    lighting2.switchOn("0xF09AC8AA/1");
    lighting2.switchOff("0xF09AC8AA/1");
</pre>

The lighting2 message controls one of three subtypes, you need to specify the
subtype to the constructor, the options are in rfxcom.defines and bound to the
Lighting2 function as Lighting2.ANSLUT, Lighting2.HOMEEASY_EU, Lighting2.AC.

Google Group
============

I've created  a Google Group for discussion of Node.js based home automation
software.

The group is nodejs-automation @ google's groups.

RfxCom events
=============

The events are named from the message identifiers sent by the RFXtrx device.

"ready"
-------
Emitted when the RFXcom has successfully opened the serial port.

"response"
----------
Emitted when a response message is received from the RFXtrx 433, sends the
status (from the RFXtrx433) and the sequence number of the message the response
is for.

"status"
--------
Emitted when a "status" message is received from the RFXtrx 433.

"elec2"
-------
Emitted when data is received from OWL electricity monitoring devices
CM119/CM160.

"security1"
-----------
Emitted when an X10 security device reports a status change.

"lighting5"
-----------
Emitted when a message is received from LightwaveRF type devices.

"th1-9"
-------
Emitted when a message is received from Oregon Scientific
Temperature/Humidity sensors.

"temp1-9"
---------
Emitted when a message is received from an Oregon Scientific temperature
sensor.

"lighting2"
-----------
Emitted when a message is received from AC/HomeEasy type devices.


changelog.md
Version 0.6.0
-------------

- Removed Lighting5 command names, replaced with Lighting5.ON instead of "On".
- Added checking for valid level in Lighting5 switchOn
- Changed the mood setting to Lighting5.MOOD1..Lighting5.MOOD5 so to switch on
  a LightwaveRF light to MOOD5 use:

    .switchOn({command: Lighting5.MOOD2})

These changes were suggested by Benjie.

- Modified the exports for Lighting2 and Lighting5, to update code use the
  subtype as a property on the function:
    lighting2 = new Lighting2(device, Lighting2.ANSLUT)

- Removed dependency on underscore.js.


Version 0.5.0
-------------

- Split the various classes out into separate files.
- Split the tests appropriately.
- Rename the LightwaveRf prototype to Lighting5. *backwards incompatible*
- Made the new Lighting5 prototype accept different types.
- Removed RfxCom.lightOn and RfxCom.lightOff because these were LightwaveRf
  specific.

Version 0.4.4
-------------

- Added support for lighting1 received messages contributed by stianeikeland.

Version 0.4.3
-------------

- Added Lighting2 class, which can control devices controlled by the lighting2
  message.

Version 0.4.2
-------------

- Fix the tamper detection for security1 devices to report status and tamper
  separately.

Version 0.4.1
-------------

- Added support for lighting2 received messages.
- Renamed signalStrength to rssi in keeping with the RFX TRX SDK document.

Version 0.4.0
-------------

- Fixed initialise to also run .open() which means you can bootstrap the device
  with one line.

Version 0.3.1
-------------

- Added .initialise to the RfxCom prototype, which handles bootstrapping the
  device.

Version 0.3.0
-------------

- Change the LightwaveRf device identifier format to 0x010203/1 which means that
  it is possible to define device as strings e.g. FRONT_LIGHT = "0x010203/1".

Version 0.2.3
-------------

- Add support for TH1-9 temperature and humidity sensors.

Version 0.2.2
-------------

- Add support for TEMP1-9 type sensors.

Version 0.2.1
-------------

- Fix bug in handler lookup and improve testing of data handler.

Version 0.2.0
-------------

- Update security1Handler event to make the types constants, and better split
  the battery-level and signal strength elements.

Version 0.1.1
-------------

- Updated the Readme.md to illustrate more use.
- Renamed the current_watts and total_watts to currentWatts and totalWatts.

Version 0.1.0
-------------
- *API CHANGE* emitted events are now a JavaScript object.
- New constants for security status messages e.g. security.MOTION.
- Refactored sending messages to the serialport through a sendMessage - still
  need to update the LightwaveRf prototype.
- Improved the test coverage.

Version 0.0.3
-------------
- Added LightwaveRf prototype for specialisation of LightwaveRf control.
  - This adds support for dimming and moodlighting control with LightwaveRF.
- Tidied up JavaScript formatting.
- Added start of rfxcom.security constants.
- Support for configuring the interface protocols.


test/lighting2.spec.js
var rfxcom = require('../lib'),
    matchers = require('./matchers'),
    FakeSerialPort = require('./helper');

describe('Lighting2 class', function () {
  var lighting2,
      fakeSerialPort,
      device;
  beforeEach(function () {
    this.addMatchers({
      toHaveSent: matchers.toHaveSent
    });
    fakeSerialPort = new FakeSerialPort();
    device = new rfxcom.RfxCom('/dev/ttyUSB0', {
      port: fakeSerialPort
    });
  });
  describe('instantiation', function () {
    it('should throw an error if no subtype is specified', function () {
      expect(function () {
        lighting2 = new rfxcom.Lighting2(device);
      }).toThrow(new Error('Must provide a subtype.'));
    });
  });
  describe('.switchOn', function () {
    beforeEach(function () {
      lighting2 = new rfxcom.Lighting2(device, rfxcom.Lighting2.ANSLUT);
    });
    it('should send the correct bytes to the serialport', function (done) {
      var sentCommandId;
      lighting2.switchOn('0x03FFFFFF/1', function (err, response, cmdId) {
        sentCommandId = cmdId;
        done();
      });
      expect(fakeSerialPort).toHaveSent([0x0B, 0x11, 0x02, 0x00, 0x03, 0xFF, 0xFF, 0xFF, 0x01, 0x01, 0x0F, 0x00]);
      expect(sentCommandId).toEqual(0);
    });
    it('should log the bytes being sent in debug mode', function (done) {
      var debugDevice = new rfxcom.RfxCom('/dev/ttyUSB0', {
        port: fakeSerialPort,
        debug: true
      }),
        debugLight = new rfxcom.Lighting2(debugDevice, rfxcom.Lighting2.ANSLUT);

      var consoleSpy = spyOn(console, 'log');
      debugLight.switchOn('0x03FFFFFF/1', done);
      expect(consoleSpy).toHaveBeenCalledWith('Sending %j', ['0B', '11', '02', '00', '03', 'FF', 'FF', 'FF', '01', '01', '0F', '00']);
    });
    it('should throw an exception with an invalid deviceId', function () {
      expect(function () {
        lighting2.switchOn('0xF09AC8');
      }).toThrow(new Error('Invalid deviceId format.'));
    });
    it('should throw an exception with an invalid deviceId', function () {
      expect(function () {
        lighting2.switchOn('0xF09AC8/1');
      }).toThrow(new Error('Invalid deviceId format.'));
    });
    it('should handle no callback', function () {
      lighting2.switchOn('0x03FFFFFF/1');
      expect(fakeSerialPort).toHaveSent([0x0b, 0x11, 2, 0, 3, 0xff, 0xff, 0xff, 1, 1, 0xf, 0]);
    });
  });
  describe('.switchOff', function () {
    beforeEach(function () {
      lighting2 = new rfxcom.Lighting2(device, rfxcom.Lighting2.ANSLUT);
    });
    it('should send the correct bytes to the serialport', function (done) {
      var sentCommandId;
      lighting2.switchOff('0x03FFFFFF/1', function (err, response, cmdId) {
        sentCommandId = cmdId;
        done();
      });
      expect(fakeSerialPort).toHaveSent([0x0b, 0x11, 2, 0, 3, 0xff, 0xff, 0xff, 1, 0, 0xf, 0]);
      expect(sentCommandId).toEqual(0);
    });
    it('should throw an exception with an invalid deviceId', function () {
      expect(function () {
        lighting2.switchOff('0xF09AC8');
      }).toThrow(new Error('Invalid deviceId format.'));
    });
    it('should handle no callback', function () {
      lighting2.switchOff('0x03FFFFFF/1');
      expect(fakeSerialPort).toHaveSent([0x0b, 0x11, 2, 0, 3, 0xff, 0xff, 0xff, 1, 0, 0xf, 0]);
    });
  });
  describe('.setLevel', function () {
    beforeEach(function () {
      lighting2 = new rfxcom.Lighting2(device, rfxcom.Lighting2.ANSLUT);
    });
    it('should send the correct bytes to the serialport', function (done) {
      var sentCommandId;
      lighting2.setLevel('0x03FFFFFF/1', 7, function (err, response, cmdId) {
        sentCommandId = cmdId;
        done();
      });
      expect(fakeSerialPort).toHaveSent([0xb, 0x11, 2, 0, 3, 0xff, 0xff, 0xff, 1, 2, 7, 0]);
      expect(sentCommandId).toEqual(0);
    });
    it('should throw an exception with an invalid level value', function () {
      expect(function () {
        lighting2.setLevel('0x03FFFFFF/1', 0x10);
      }).toThrow(new Error('Invalid level value must be in range 0-15.'));
    });
    it('should throw an exception with an invalid deviceId', function () {
      expect(function () {
        lighting2.setLevel('0xF09AC8', 3);
      }).toThrow(new Error('Invalid deviceId format.'));
    });
    it('should handle no callback', function () {
      lighting2.setLevel('0x03FFFFFF/1', 5);
      expect(fakeSerialPort).toHaveSent([0x0b, 0x11, 2, 0, 3, 0xff, 0xff, 0xff, 1, 2, 5, 0]);
    });
  });
});


test/lighting5.spec.js
var rfxcom = require('../lib'),
    matchers = require('./matchers'),
    FakeSerialPort = require('./helper');

  beforeEach(function (){
    this.addMatchers({
      toHaveSent: matchers.toHaveSent
    });
    fakeSerialPort = new FakeSerialPort();
    device = new rfxcom.RfxCom('/dev/ttyUSB0', {
      port: fakeSerialPort
    });
  });

describe('Lighting5 class', function(){
  var lighting5,
      fakeSerialPort,
      device;
  beforeEach(function(){
    this.addMatchers({
      toHaveSent: matchers.toHaveSent
    });
    fakeSerialPort = new FakeSerialPort();
    device = new rfxcom.RfxCom('/dev/ttyUSB0', {
      port: fakeSerialPort
    });
  });
  describe('instantiation', function () {
    it('should throw an error if no subtype is specified', function () {
      expect(function () {
        lighting5 = new rfxcom.Lighting5(device);
      }).toThrow(new Error('Must provide a subtype.'));
    });
  });
  describe('.switchOn', function(){
    beforeEach(function (){
      lighting5 = new rfxcom.Lighting5(device, rfxcom.Lighting5.LIGHTWAVERF);
    });
    it('should send the correct bytes to the serialport', function(done){
      var sentCommandId;
          lighting5.switchOn('0xF09AC8/1', function(err, response, cmdId){
            sentCommandId = cmdId;
            done();
          });
      expect(fakeSerialPort).toHaveSent([10, 20, 0, 0, 0xF0, 0x9A, 0xC8, 1, 1, 0x1f, 0]);
      expect(sentCommandId).toEqual(0);
    });
    it('should throw an exception with an invalid deviceId', function(){
      expect(function(){
        lighting5.switchOn('0xF09AC8');
      }).toThrow(new Error('Invalid deviceId format.'));
    });
    it('should handle mood lighting', function(done){
      lighting5.switchOn('0xF09AC8/1', {
        command: rfxcom.Lighting5.MOOD2
      }, function(){
        done();
      });
      expect(fakeSerialPort).toHaveSent([10, 20, 0, 0, 0xF0, 0x9A, 0xC8, 1, 4, 0x1f, 0]);
    });
    it('should throw an exception with an invalid level value', function(){
      expect(function(){
        lighting5.switchOn('0xF09AC8/1', {
          level: 32
        });
      }).toThrow(new Error('Invalid level value must be in range 1-31.'));
    });
    it('should send the level if one is specified', function(done){
      lighting5.switchOn('0xF09AC8/1', {
        level: 0x14,
      }, function(){
        done();
      });
      expect(fakeSerialPort).toHaveSent([10, 20, 0, 0, 0xF0, 0x9A, 0xC8, 1, 1, 0x14, 0]);
    });
    it('should handle no callback', function(){
      lighting5.switchOn('0xF09AC8/1', {
        level: 0x10
      });
      expect(fakeSerialPort).toHaveSent([10, 20, 0, 0, 0xF0, 0x9A, 0xC8, 1, 1, 0x10, 0]);
    });
  });
  describe('.switchOff', function(){
    beforeEach(function (){
      lighting5 = new rfxcom.Lighting5(device, rfxcom.Lighting5.EMW100);
    });
    it('should send the correct bytes to the serialport', function(done){
      var sentCommandId;
          lighting5.switchOff('0xF09AC8/1', function(err, response, cmdId){
            sentCommandId = cmdId;
            done();
          });
      expect(fakeSerialPort).toHaveSent([10, 20, 1, 0, 0xF0, 0x9A, 0xC8, 1, 0, 0x1f, 0]);
      expect(sentCommandId).toEqual(0);
    });
    it('should handle no callback', function(){
      lighting5.switchOff('0xF09AC8/1');
      expect(fakeSerialPort).toHaveSent([10, 20, 1, 0, 0xF0, 0x9A, 0xC8, 1, 0, 0x1f, 0]);
    });
  });
});



lib/lighting2.js
module.exports = Lighting2;
const defines = require('./defines');

/*
 * This is a class for controlling Lighting2 lights.
 */
function Lighting2(rfxcom, subtype) {
  var self = this;

  self.rfxcom = rfxcom;
  self.subtype = subtype;

  if (typeof self.subtype === "undefined") {
    throw new Error("Must provide a subtype.");
  };
}

/*
 * Puts the defines onto the Lighting2 function
 * Allows access to the message types through
 * Lighting2.ANSLUT etc.
 */
for (var k in defines.lighting2) {
  Lighting2[k] = defines.lighting2[k];
}

/*
 * Splits the device id x/y and returns the components, the deviceId will be
 * returned as the component bytes, ready for sending.
 *
 * Throws an Error if the format is invalid or if the deviceId is not the
 * correct length.
 */
Lighting2.prototype._splitDeviceId = function (deviceId) {
  var parts = deviceId.split("/"),
      idBytes;
  if (parts.length !== 2) {
    throw new Error("Invalid deviceId format.");
  };
  idBytes = this.rfxcom.stringToBytes(parts[0]);
  if (idBytes.length !== 4) {
    throw new Error("Invalid deviceId format.");
  };
  return {
    idBytes: idBytes,
    unitCode: parts[1],
  };
};

Lighting2.prototype._sendCommand = function (deviceId, command, level, callback) {
  var self = this,
      device = self._splitDeviceId(deviceId),
      cmdId = self.rfxcom.getCmdNumber(),
      level = level || 0xf,
      buffer = [0x0b, defines.messageTypes.LIGHTING2, self.subtype, cmdId, device.idBytes[0],
                device.idBytes[1], device.idBytes[2], device.idBytes[3],
                device.unitCode, command, level, 0];

  if (self.rfxcom.options.debug) {
    console.log("Sending %j", self.rfxcom.dumpHex(buffer));
  }
  self.rfxcom.serialport.write(buffer, function (err, response) {
    if (typeof callback === "function") {
      callback(err, response, cmdId);
    }
  });
  return cmdId;
};

/*
 * Switch on deviceId/unitCode
 */
Lighting2.prototype.switchOn = function (deviceId, callback) {
  return this._sendCommand(deviceId, 1, undefined, callback);
};

/*
 * Switch off deviceId/unitCode
 */
Lighting2.prototype.switchOff = function (deviceId, callback) {
  return this._sendCommand(deviceId, 0, undefined, callback);
};

/*
 * Switch off deviceId/unitCode
 */
Lighting2.prototype.setLevel = function (deviceId, level, callback) {
  if ((level < 0) || (level > 0xf)) {
    throw new Error("Invalid level value must be in range 0-15.");
  };
  return this._sendCommand(deviceId, 2, level, callback);
};


lib/lighting5.js
module.exports = Lighting5;
const defines = require('./defines');

/*
 * This is a class for controlling LightwaveRF lights.
 */
function Lighting5(rfxcom, subtype) {
  var self = this;

  self.rfxcom = rfxcom;
  self.subtype = subtype;

  if (typeof self.subtype === "undefined") {
    throw new Error("Must provide a subtype.");
  };
}

/*
 * Puts the defines onto the Lighting5 function
 * Allows access to the message types through
 * Lighting5.LIGHTWAVERF etc.
 */
for (var k in defines.lighting5) {
  Lighting5[k] = defines.lighting5[k];
}

/*
 * Splits the device id x/y and returns the components, the deviceId will be
 * returned as the component bytes, ready for sending.
 *
 * Throws an Error if the format is invalid.
 */
Lighting5.prototype._splitDeviceId = function (deviceId) {
  var parts = deviceId.split("/");
  if (parts.length !== 2) {
    throw new Error("Invalid deviceId format.");
  };
  return {
    idBytes: this.rfxcom.stringToBytes(parts[0]),
    unitCode: parts[1],
  };
};


Lighting5.prototype._sendCommand = function (deviceId, command, level, callback) {
  var self = this,
      device = self._splitDeviceId(deviceId),
      cmdId = self.rfxcom.getCmdNumber(),
      level = level || 0x1f,
      buffer = [0x0a, defines.messageTypes.LIGHTING5, self.subtype, cmdId,
                device.idBytes[0], device.idBytes[1], device.idBytes[2], 
                device.unitCode, command, level, 0];

  self.rfxcom.serialport.write(buffer, function (err, response) {
    if (typeof callback === "function") {
        callback(err, response, cmdId);
    }
  });
  return cmdId;
};


/*
 * Switch on deviceId/unitCode
 */
Lighting5.prototype.switchOn = function (deviceId, options, callback) {
  if (typeof options === "function") {
    callback = options;
    options = {};
  };
  if (typeof options === "undefined") {
    options = {};
  };
  if (typeof options.level !== "undefined") {
    if (options.level < 1 || options.level > 31) {
      throw new Error("Invalid level value must be in range 1-31.");
      }
  };
  var command = options.command || Lighting5.ON,
      level = options.level || 0;

  return this._sendCommand(deviceId, command, level, callback);
};


/*
 * Switch off deviceId/unitCode
 */
Lighting5.prototype.switchOff = function (deviceId, callback) {
  return this._sendCommand(deviceId, 0, undefined, callback);
};


lib/rfxcom.js
var serialport = require("serialport"),
    EventEmitter = require("events").EventEmitter,
    util = require("util");

function RfxCom(device, options) {
  var self = this;

  self.options = options || {};
  self.handlers = {
    0x01: "statusHandler",
    0x02: "messageHandler",
    0x10: "lighting1Handler",
    0x11: "lighting2Handler",
    0x14: "lighting5Handler",
    0x5a: "elec2Handler",
    0x20: "security1Handler",
    0x50: "temp19Handler",
    0x52: "temphumidity19Handler"
  };

  // Running counter for command numbers.
  self._cmd = 0;

  // Allow for faking out the SerialPort
  if (typeof self.options.port !== "undefined") {
      self.serialport = options.port;
  }

  // Store the device to use
  self.device = device;

  // This is a buffering parser which accumulates bytes until it receives the
  // number of bytes specified in the first byte of the message.
  // It relies on a flushed buffer, to ensure the first byte corresponds to the
  // size of the first message.
  // The 'data' message emitted has all the bytes from the message.
  self.rfxtrxParser = function () {
    var data = [],
        requiredBytes = 0;
    return function (emitter, buffer) {
      // Collect data
      data.push.apply(data, buffer);
      if (requiredBytes === 0) {
        requiredBytes = buffer[0] + 1;
      }
      if (data.length >= requiredBytes) {
        emitter.emit("data", data.slice(0, requiredBytes + 1));
        data = data.slice(requiredBytes);
        requiredBytes = 0;
      }
    };
  };
}
util.inherits(RfxCom, EventEmitter);

RfxCom.prototype.open = function () {
  var self = this;

  if (typeof self.serialport === "undefined") {
    self.serialport = new serialport.SerialPort(self.device, {
      baudrate : 38400,
      parser: self.rfxtrxParser()
    });
  }
  self.serialport.on("open", function () {
    self.emit("ready");
  });

  // Add data read event listener
  self.serialport.on("data", function (data) {
    if (self.options.debug) {
      console.log("Received: %s", self.dumpHex(data));
    }

    var length = data.shift(),
        packetType = data.shift(),
        handler = self.handlers[packetType];

    if (typeof handler !== "undefined") {
      self[handler](data);
    } else {
      if (self.options.debug) {
        console.log("Unhandled packet type = %s", self.dumpHex([packetType]));
      }
    }
  });

  self.serialport.on("error", function (msg) {
    console.log("Error %s", msg);
  });

  self.serialport.on("end", function () {
    console.log("Received 'end'");
  });

  self.serialport.on("drain", function () {
    console.log("Received 'drain'");
  });
};

RfxCom.prototype.messageHandler = function (data) {
  var self = this,
      subtype = data[0],
      seqnbr = data[1],
      responses = {
        0: "ACK - transmit OK",
        1: "ACK - transmit delayed",
        2: "NAK - transmitter did not lock onto frequency",
        3: "NAK - AC address not allowed"
      },
      message = data[2];
  self.emit("response", responses[message], seqnbr);
};

RfxCom.prototype.initialise = function (callback) {
  var self = this;

  self.on("ready", function () {
    self.reset(function () {
      self.delay(500);
      self.flush();
      self.getStatus(callback);
    });
  });
  self.open();
};


/**
 *
 * Called by the data event handler when data arrives from the device with
 * information about its settings.
 *
 */
RfxCom.prototype.statusHandler = function (data) {
  var self = this,
      receiverTypes = {
        0x50: "310MHz",
        0x51: "315MHz",
        0x52: "433.92MHz receiver only",
        0x53: "433.92MHz transceiver",
        0x55: "868.00MHz",
        0x56: "868.00MHz FSK",
        0x57: "868.30MHz",
        0x58: "868.30MHz FSK",
        0x59: "868.35MHz",
        0x5A: "868.35MHz FSK",
        0x5B: "868.95MHz"
        },
        subtype = data[0],
        seqnbr = data[1],
        cmnd = data[2],
        receiverType = receiverTypes[data[3]],
        firmwareVersion = data[4],
        evt = {
          subtype: subtype,
          seqnbr: seqnbr,
          cmnd: cmnd,
          receiverType: receiverType,
          firmwareVersion: firmwareVersion
        };
  self.emit("status", evt);
};

/**
 *
 * Fetches a "command number" sequence number for identifying requests sent to
 * the device.
 *
 */
RfxCom.prototype.getCmdNumber = function () {
  var self = this;

  if (self._cmd > 256) {
    self._cmd = 0;
  }
  return self._cmd++;
};

/**
 *
 * Internal function for sending messages to the device.
 *
 */
RfxCom.prototype.sendMessage = function (type, subtype, cmd, extra, callback) {
  var self = this,
      cmdId = this.getCmdNumber(),
      byteCount = extra.length + 4,
      buffer = [byteCount, type, subtype, cmdId, cmd];

  buffer = buffer.concat(extra);
  self.serialport.write(buffer, callback);
  return cmdId;
};


/**
 *
 * Writes the reset sequence to the RFxtrx433.
 *
 */
RfxCom.prototype.reset = function (callback) {
  var self = this;
  return self.sendMessage(0, 0, 0, [0, 0, 0, 0, 0, 0, 0, 0, 0], callback);
};

/**
 *
 * Calls flush on the underlying SerialPort.
 *
 */
RfxCom.prototype.flush = function (callback) {
  var self = this;
  self.serialport.flush(callback);
};


/*
 * Sends the getstatus bytes to the interface.
 */
RfxCom.prototype.getStatus = function (callback) {
  var self = this;
  return self.sendMessage(0, 0, 2, [0, 0, 0, 0, 0, 0, 0, 0, 0], callback);
};

/*
 * Enables reception of different protocols.
 */
RfxCom.prototype.enable = function (protocols, callback) {
  var self = this,
      cmdId = this.getCmdNumber(),
      msg = {},
      buffer;

  protocols.forEach(function (protocol) {
    if (typeof msg[protocol.msg] === "undefined") {
      msg[protocol.msg] = protocol.bit;
    } else {
      msg[protocol.msg] |= protocol.bit;
    }
  });

  buffer = [0x0d, 0, 0, cmdId, 3, 0x53, 0, 0, msg["4"], msg["5"], 0, 0, 0, 0];

  self.serialport.write(buffer, function (err, response) {
    if (callback) {
      return callback(err, response, cmdId);
    }
  });
  return cmdId;
};

/**
 *
 * Wait for at least the specified number of milliseconds.
 *
 * This may not wait precisely for that number of milliseconds due to the
 * vagaries of system scheduling, but no fewer than the specified ms.
 *
 */
RfxCom.prototype.delay = function (ms) {
    ms += +new Date();
    while (+new Date() < ms) {
      // Do nothing...
    }
};

RfxCom.prototype.dumpHex = function (buffer, prefix) {
  prefix = prefix || "";
  function dec2hex(value) {
    var hexDigits = "0123456789ABCDEF";
    return prefix + (hexDigits[value >> 4] + hexDigits[value & 15]);
  }
  return buffer.map(dec2hex);
};

/**
 *
 * Converts an array of 4 bytes to a 32bit integer.
 *
 */
RfxCom.prototype.bytesToUint32 = function (bytes) {
  return (bytes[0] * Math.pow(2, 24)) + (bytes[1] << 16) +
         (bytes[2] << 8) + bytes[3];
};

/**
 *
 * Converts an array of 6 bytes to a 48bit integer.
 *
 */
RfxCom.prototype.bytesToUint48 = function (bytes) {
  return (bytes[0] * Math.pow(2, 40)) + (bytes[1] * Math.pow(2, 32)) +
         (bytes[2] * Math.pow(2, 24)) + (bytes[3] << 16) + (bytes[4] << 8) +
         bytes[4];
};

/**
 *
 * Converts a string to an array of Byte values.
 *
 * e.g. stringToBytes("202020") == [32, 32, 32]
 *
 */
RfxCom.prototype.stringToBytes = function (bytes) {
  var result = [],
      byteString,
      value,
      i;
  for (i = 0; i < bytes.length; i += 2) {
    byteString = bytes.substr(i, 2);
    if (byteString !== "0x") {
      value = parseInt(bytes.substr(i, 2), 16);
      if (typeof value !== "undefined") {
        result.push(value);
      }
    }
  }
  return result;
};

/**
 *
 * Called by the data event handler when data arrives from an OWL CM119/CM160
 * power measurement device.
 *
 * Calculates the current usage and total usage from the bytes sent, and emits
 * an "elec2" message for handling.
 *
 */
RfxCom.prototype.elec2Handler = function (data) {
  var self = this,
      subtypes = {
          1: "CM119/160"
      },
      TOTAL_DIVISOR = 223.666,
      subtype = subtypes[data[0]],
      seqnbr = data[1],
      id = "0x" + self.dumpHex(data.slice(2, 4), false).join(""),
      count = data[4],
      instant = data.slice(5, 9),
      total = data.slice(9, 15),
      currentWatts = self.bytesToUint32(instant),
      totalWatts = self.bytesToUint48(total) / TOTAL_DIVISOR,
      roundedTotal = Math.round(totalWatts * Math.pow(10, 2)) / Math.pow(10, 2),
      evt =  {
        subtype: subtype,
        id: id,
        currentWatts: currentWatts,
        totalWatts: roundedTotal
      };

    self.emit("elec2", evt);
};

/**
 *
 * Called by the data event handler when data arrives from various security
 * devices.
 *
 */
RfxCom.prototype.security1Handler = function (data) {
    var self = this,
        subtype = data[0],
        seqnbr = data[1],
        id = "0x" + self.dumpHex(data.slice(2, 5), false).join(""),
        deviceStatus = data[5] & ~0x80,
        batterySignalLevel = data[6],
        evt = {
          subtype: subtype,
          id: id,
          deviceStatus: deviceStatus,
          batteryLevel: batterySignalLevel >> 4,
          rssi: batterySignalLevel & 0x0f,
          tampered: data[5] & 0x80
        };

    self.emit("security1", evt);
};

/**
 *
 * Called by the data event handler when data arrives from temp1-9
 * devices.
 *
 */
RfxCom.prototype.temp19Handler = function (data) {
  var self = this,
      subtype = data[0],
      seqnbr = data[1],
      id = "0x" + self.dumpHex(data.slice(2, 4), false).join(""),
      temperature = ((data[4] & 0x7f) * 256 + data[5]) / 10,
      signbit = data[4] & 0x80,
      batterySignalLevel = data[6],
      evt = {
        subtype: subtype,
        id: id,
        seqnbr: seqnbr,
        temperature: temperature * (signbit ? -1 : 1),
        batteryLevel: batterySignalLevel >> 4,
        rssi: batterySignalLevel & 0x0f,
      };
  self.emit("temp" + subtype, evt);
};

/**
 *
 * Called by the data event handler when data arrives from th1-9
 * devices.
 *
 */
RfxCom.prototype.temphumidity19Handler = function (data) {
  var self = this,
      subtype = data[0],
      seqnbr = data[1],
      id = "0x" + self.dumpHex(data.slice(2, 4), false).join(""),
      temperature = ((data[4] & 0x7f) * 256 + data[5]) / 10,
      signbit = data[4] & 0x80,
      humidity = data[6],
      humidityStatus = data[7],
      batterySignalLevel = data[8],
      evt = {
        subtype: subtype,
        id: id,
        seqnbr: seqnbr,
        temperature: temperature * (signbit ? -1 : 1),
        humidity: humidity,
        humidityStatus: humidityStatus,
        batteryLevel: batterySignalLevel >> 4,
        rssi: batterySignalLevel & 0x0f,
      };

  self.emit("th" + subtype, evt);
};

/**
 *
 * Called by the data event handler when data arrives from a Lighting1
 * light control device.
 *
 */
RfxCom.prototype.lighting1Handler = function (data) {
  var self = this,
      subtypes = {
          0: "X10",
          1: "ARC",
          2: "ELRO AB400D",
          3: "Waveman",
          4: "EMW200",
          5: "Impuls",
          6: "RisingSun",
          7: "Philips SBC"
      },
      commands = {
          0: "Off",
          1: "On",
          5: "All Off",
          6: "All On",
          7: "Chime"
      },
      subtype = subtypes[data[0]],
      seqnbr = data[1],
      housecode = String.fromCharCode(data[2]),
      unitcode = data[3],
      command = commands[data[4]],
      rssi = data[5] >> 4,
      id = self.dumpHex(data.slice(2,4), false).join(""),
      evt;

  evt = {
    id: "0x" + id,
    subtype: subtype,
    seqnbr: seqnbr,
    housecode: housecode,
    unitcode: unitcode,
    command: command,
    rssi: rssi
  };

  self.emit("lighting1", evt);
};

/**
 *
 * Called by the data event handler when data arrives from a LightwaveRF/Siemens
 * light control device.
 *
 */
RfxCom.prototype.lighting2Handler = function (data) {
  var self = this,
      subtypes = {
          0: "AC",
          1: "HomeEasy EU",
          2: "ANSLUT"
      },
      commands = {
        0: "Off",
        1: "On",
        2: "Set Level",
        3: "Group Off",
        4: "Group On",
        5: "Set Group Level"
      },
      subtype = subtypes[data[0]],
      seqnbr = data[1],
      idBytes = data.slice(2, 6),
      unitcode = data[6],
      command = commands[data[7]],
      level = data[8],
      rssi = data[9] & 0x0f,
      evt;

  idBytes[0] &= ~0xfc; // "id1 : 2"
  evt = {
    subtype: subtype,
    seqnbr: seqnbr,
    id: "0x" + self.dumpHex(idBytes, false).join(""),
    unitcode: unitcode,
    command: command,
    level: level,
    rssi: rssi
  };

  self.emit("lighting2", evt);
};

/**
 *
 * Called by the data event handler when data arrives from a LightwaveRF/Siemens
 * light control device.
 *
 */
RfxCom.prototype.lighting5Handler = function (data) {
  var self = this,
      subtypes = {
          0: "LightwaveRF, Siemens",
          1: "EMW100 GAO/Everflourish"
      },
      subtype = subtypes[data[0]],
      seqnbr = data[1],
      id = "0x" + self.dumpHex(data.slice(2, 5), false).join(""),
      evt = {
        subtype: subtype,
        seqnbr: data[1],
        id: id,
        unitcode: data[5],
        command: data[6]
      };
  self.emit("lighting5", evt);
};

module.exports = RfxCom;
